### ISSUE TITLE
Enable user to understand when resource is reconciled

### ISSUE THREAD

**Issue by pedjak** [crossplane/crossplane#4655]
### What problem are you facing?

After a claim/composition update, user is not able to figure out if the resource was reconciled. Monitoring sync and ready statuses is not helpful, if the update went well.

### How could Crossplane help solve your problem?

It is considered as [a good practice for k8s operators](https://alenkacz.medium.com/kubernetes-operator-best-practices-implementing-observedgeneration-250728868792) to emit object's [`metadata.generation`](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#metadata) in `status.observedGeneration`.

The field value is a sequence number representing a specific generation of the desired state. Set by the system and monotonically increasing, per-resource. It gets increased when object `.spec` changes. For example, at the first update, and before the claim got reconciled, user would see the following:

```yaml
apiVersion: nop.example.org/v1alpha1
kind: NopResource
metadata:
  name: claim
  generation: 2
status:
  observedGeneration: 1
```

After the reconciliation, `obeservedGeneration` becomes `2`. Hence, user is able to detect if the process takes too long or simply not moving forward.

In case of claims, given that their labels/annotations are propagated to the composite, it might be needed to add `status.observedLabels` and/or `status.observedAnnotations` as well, because `generation` value is not changed on metadata updates.

**Comment by negz** [comment]
Yeah, let's add `observedGeneration` to all the things. Cross-ref to https://github.com/crossplane/crossplane/issues/2672 where this was briefly mentioned.

**Comment by negz** [comment]
/fresh

We merged a change that added this field to various resource schemas, but there's not yet any code that actually sets it.

**Comment by MisterMX** [comment]
As an addition to this we would like to transport the generation of the claim to the composite and back to the claim again to validate with which version of the claim the composed resources were created.

Since most errors happen during the compose phase which is done during the composite reconcile, it is only of minimal benefit to users to know that a claim has been reconciled successfully. The real update users are interested in is that of the external resources.

Since we don't have access to the claim resource itself during the compose phase (i.e. in functions), we cannot do this with a custom implementation and have to rely on Crossplane to fix it.

(if Crossplane does for whatever reason not want to do this, it should be sufficient for us to have the claim send with the function request so we can implement it on our own)

**Comment by n3wscott** [comment]
I think this is done. I have made the change in #6420 and based on my testing, core crossplane resources report the correct observed generation, though the ecosystem will need to be caught up when it comes to MRs and conditions that are owned by a Provider.

As Claims are deprecated, I did not implement any new fancy label propagation. 

Please re-open if there is further work to report.
